File: src/main/java/org/metaborg/spoofax/intellij/languages/SpoofaxFileType.java
package org.metaborg.spoofax.intellij.languages;

import com.intellij.lang.Language;
import com.intellij.openapi.fileTypes.LanguageFileType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.metaborg.core.language.ILanguage;
import org.metaborg.spoofax.intellij.idea.model.SpoofaxIcons;
import org.metaborg.spoofax.intellij.logging.InjectLogger;
import org.slf4j.Logger;

import javax.swing.*;
import java.util.HashSet;
import java.util.Set;

/**
 * Base class for all Spoofax file types.
 */
public abstract class SpoofaxFileType extends LanguageFileType {

    /**
     * Initializes a new instance of the {@link SpoofaxFileType} class.
     *
     * @param language The language.
     */
    protected SpoofaxFileType(@NotNull final SpoofaxIdeaLanguage language) {
        super(language);
    }

    /**
     * Gets the Spoofax language.
     *
     * @return The Spoofax language.
     */
    public final ILanguage getSpoofaxLanguage() {
        return ((SpoofaxIdeaLanguage)super.getLanguage()).language();
    }

    @NotNull
    @Override
    public String getName() {
        return this.getSpoofaxLanguage().name();
    }

    @NotNull
    @Override
    public String getDescription() {
        // TODO:
        return "A spoofax language";
    }

    @NotNull
    @Override
    public String getDefaultExtension() {
        return LanguageUtils.getDefaultExtension(this.getSpoofaxLanguage().activeImpl());
    }

    @NotNull
    public Set<String> getExtensions() {
        return LanguageUtils.getExtensions(this.getSpoofaxLanguage().activeImpl());
    }

    @Nullable
    @Override
    public Icon getIcon() {
        return SpoofaxIcons.INSTANCE.Default;
    }

}

File: org.metaborg.intellij/src/main/kotlin/org/metaborg/intellij/idea/files/SpoofaxFileType.kt
package org.metaborg.intellij.idea.files

import com.google.inject.Inject
import com.google.inject.assistedinject.Assisted
import com.intellij.openapi.fileTypes.LanguageFileType
import com.virtlink.editorservices.intellij.files.IFileTypeExt
import org.metaborg.intellij.idea.graphics.IIconManager
import org.metaborg.intellij.idea.languages.SpoofaxIdeaLanguage
import org.metaborg.intellij.idea.languages.defaultExtension
import org.metaborg.intellij.idea.languages.extensions
import javax.swing.Icon

/**
 * The file type of a Spoofax file.
 *
 * Every Spoofax language gets its own [SpoofaxFileType].
 */
class SpoofaxFileType @Inject constructor(
        @Assisted language: SpoofaxIdeaLanguage,
        private val iconManager: IIconManager)
    : LanguageFileType(language), IFileTypeExt {

    companion object {
        /**
         * The dummy extension that is used when the default extension could not be determined.
         *
         * Should not have a leading dot.
         */
        const val DUMMY_EXTENSION = "spoofax"
    }

    interface IFactory {
        fun create(language: SpoofaxIdeaLanguage)
                : SpoofaxFileType
    }

    /**
     * Gets the Spoofax language of this file type.
     */
    val spoofaxLanguage: SpoofaxIdeaLanguage get() = super.getLanguage() as SpoofaxIdeaLanguage

    override fun getName(): String {
        // NOTE: This is the ID of the language.
        return this.spoofaxLanguage.language.name()
    }

    override fun getDescription(): String {
        // NOTE: This is displayed as the name of the language.
        // TODO SPOOFAX: This should be a human-readable custom name, a 'displayName'.
        return "$name (Spoofax)"
    }

    override fun getIcon(): Icon? {
        // TODO SPOOFAX: Spoofax should allow a custom icon for the language.
        return this.iconManager.getLanguageFileIcon(this.spoofaxLanguage.language)
    }

    override fun getDefaultExtension(): String {
        return this.spoofaxLanguage.language.defaultExtension ?: DUMMY_EXTENSION
    }

    override val extensions: Set<String>
        get() = this.spoofaxLanguage.language.extensions
}